libs/wire-api/src/Data/Swagger/Typed.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Data.Swagger.Typed
+  ( Schema,
+    NamedSchema,
+    ToTypedSchema (..),
+    TypedSchema (TypedSchema),
+    field,
+    field',
+    named,
+    unnamed,
+    array,
+    enum,
+    schema,
+    untypedSchema,
+  )
+where
+
+import Control.Applicative
+import Control.Lens (at, view, (.~), (?~))
+import Data.Aeson (Value)
+import Data.Proxy
+import qualified Data.Swagger as S
+import qualified Data.Swagger.Declare as S
+import qualified Data.Swagger.Internal.Schema as S
+import Imports
+
+type Declare = S.Declare (S.Definitions S.Schema)
+
+-- | A typed Swagger schema
+newtype Schema (a :: *) = Schema (Declare S.Schema)
+  deriving (Functor)
+
+-- FUTUREWORK: store enough information in 'Schema' to synthesise
+-- ToJSON and FromJSON instances
+
+data NamedSchema a = NamedSchema
+  { _nsSchema :: Declare S.NamedSchema,
+    _nsDescription :: Maybe Text
```

Comment:
Not sure how, because the `Maybe` is important for the `Applicative` and `Alternative` instances. One could argue that those instances shouldn't be there, but they are convenient in certain cases (e.g. [here](https://github.com/wireapp/wire-server/pull/1461/commits/53a2992e8773451a6940d8808390a454b4202a57#diff-30064c2e38bdfba879d773c11abc7e226ab8ae4d3c7435db650a6a4d95f7fc9dR284)).

Also, if we eventually want this to make it upstream, it should support optional descriptions.